server/catgenome/src/main/java/com/epam/catgenome/util/SimpleMotifSearchIterator.java

```diff
@@ -0,0 +1,179 @@
+
+public class SimpleMotifSearchIterator implements Iterator<Motif>  {
+
+    private final Deque<Match> positiveMatches;
+    private final Deque<Match> negativeMatches;
+    private final String contig;
+    private final byte[] sequence;
+    private final int offset;
+    private final boolean includeSequence;
+
+    public SimpleMotifSearchIterator(final byte[] seq, final String regex,
+                                     final StrandSerializable strand, final String contig,
+                                     final int start, final boolean includeSequence) {
+        if (strand != null && strand != StrandSerializable.POSITIVE && strand != StrandSerializable.NEGATIVE) {
+            throw new IllegalStateException("Not supported strand: " + strand);
+        }
+        this.contig = contig;
+        this.sequence = seq;
+        this.offset = start;
+        this.includeSequence = includeSequence;
+
+        final String iupacRegex = MotifSearcher.convertIupacToRegex(regex);
+        final String invertedRegex = invertCurrentRegex(regex.toUpperCase(Locale.ROOT));
+        final Pattern patternPositive = Pattern.compile(iupacRegex, Pattern.CASE_INSENSITIVE);
+        final Pattern patternNegative = Pattern.compile(invertedRegex, Pattern.CASE_INSENSITIVE);
+        if (strand == null) {
+            this.positiveMatches = populatePositiveMatches(patternPositive.matcher(new String(seq)));
+            this.negativeMatches = populatePositiveMatches(patternNegative.matcher(new String(seq)));
+        } else if (strand == StrandSerializable.POSITIVE) {
+            this.positiveMatches = populatePositiveMatches(patternPositive.matcher(new String(seq)));
+            this.negativeMatches = new LinkedList<>();
+        } else {
+            this.positiveMatches = new LinkedList<>();
+            this.negativeMatches = populatePositiveMatches(patternNegative.matcher(new String(seq)));
+        }
+    }
+
+    @Override
+    public boolean hasNext() {
+        return !positiveMatches.isEmpty() || !negativeMatches.isEmpty();
+    }
+
+    @Override
+    public Motif next() {
+        Match match;
+        StrandSerializable currentStrand;
+        if (negativeMatches.isEmpty()) {
+            match = positiveMatches.removeFirst();
+            currentStrand = StrandSerializable.POSITIVE;
+        } else if (positiveMatches.isEmpty() || negativeMatches.peekLast().start < positiveMatches.peekFirst().start) {
+            match = negativeMatches.removeLast();
+            currentStrand = StrandSerializable.NEGATIVE;
+        } else {
+            match = positiveMatches.removeFirst();
+            currentStrand = StrandSerializable.POSITIVE;
+        }
+        if (includeSequence) {
+            return getMotif(contig, match.start, match.end, currentStrand);
+        }
+        return new Motif(contig, match.start + offset, match.end + offset, currentStrand, null);
+    }
+
+    private String invertCurrentRegex(final String regex) {
```

Comment:
better  to move logic of `invertCurrentRegex` and `checkAndRevert` to separate util class, to have clean logic here